source/common/network/lc_trie.h

```diff
@@ -0,0 +1,567 @@
+      ComputePair compute(0, 0);
+
+      // Compute the new prefix for the range between ip_prefixes_[first] and
+      // ip_prefixes_[first + n - 1].
+      IpType high = removeBits<IpType, address_size>(prefix, ip_prefixes_[first].ip_);
+      IpType low = removeBits<IpType, address_size>(prefix, ip_prefixes_[first + n - 1].ip_);
+      uint32_t index = prefix;
+
+      // Find the index at which low and high diverge to get the skip.
+      while (extractBits<IpType, address_size>(index, 1, low) ==
+             extractBits<IpType, address_size>(index, 1, high)) {
+        ++index;
+      }
+      compute.prefix_ = index;
+
+      // For 2 elements, use a branching factor of 2(2^1).
+      if (n == 2) {
+        compute.branch_ = 1;
+        return compute;
+      }
+
+      // According to the original LC-Trie paper, a large branching factor(suggested value: 16)
+      // at the root increases performance.
+      if (root_branching_factor_ > 0 && prefix == 0 && first == 0) {
+        compute.branch_ = root_branching_factor_;
+        return compute;
+      }
+
+      // Compute the number of bits required for branching by checking all patterns in the set are
+      // covered. Ex (b=2 {00, 01, 10, 11}; b=3 {000,001,010,011,100,101,110,111}, etc)
+      uint32_t branch = 1;
+      uint32_t count;
+      do {
+        ++branch;
+
+        // Check if the current branch factor with the fill factor can contain all of the nodes
+        // in the current range or if the current branching factor is larger than the
+        // IP address_size.
+        if (n < fill_factor_ * (1 << branch) ||
+            static_cast<uint32_t>(compute.prefix_ + branch) > address_size) {
+          break;
+        }
+
+        // Start by checking the bit patterns at ip_prefixes_[first] through
+        // ip_prefixes_[first + n-1].
+        index = first;
+        // Pattern to search for.
+        uint32_t pattern = 0;
+        // Number of patterns found while looping through the list.
+        count = 0;
+
+        // Search for all patterns(values) within 1<<branch.
+        while (pattern < static_cast<uint32_t>(1 << branch)) {
+          bool pattern_found = false;
+          // Keep on looping until either all nodes in the range have been visited or
+          // an IP prefix doesn't match the pattern.
+          while (index < first + n &&
+                 static_cast<uint32_t>(extractBits<IpType, address_size>(
+                     compute.prefix_, branch, ip_prefixes_[index].ip_)) == pattern) {
+            ++index;
+            pattern_found = true;
+          }
+
+          if (pattern_found) {
+            ++count;
+          }
+          ++pattern;
+        }
+        // Stop iterating once the size of the branch (with the fill factor ratio) is able to
+        // contain all of the prefixes within the current range of ip_prefixes_[first] to
+        // ip_prefixes_[first+n-1].
+      } while (count >= fill_factor_ * (1 << branch));
```

Comment:
I am a little concerned about the default fill factor of 0. That would mean that any count would keep increasing the branch going up and collapsing the trie

Reply:
fill factor will validated in the config to have this range (0, 1]